Applications/spire/source/charting/chart_view.cpp

```diff
@@ -33,96 +45,140 @@ ChartView::ChartView(ChartValue::Type x_axis_type, ChartValue::Type y_axis_type,
       m_y_axis_type(y_axis_type),
       m_timestamp_format("%H:%M:%S"),
       m_label_font("Roboto"),
-      m_item_delegate(new CustomVariantItemDelegate(this)) {
+      m_font_metrics(QFont()),
+      m_item_delegate(new CustomVariantItemDelegate(this)),
+      m_dashed_line_pen(Qt::white, scale_width(1), Qt::CustomDashLine),
+      m_label_text_color(QColor("#25212E")) {
   setFocusPolicy(Qt::NoFocus);
+  setMouseTracking(true);
+  setAttribute(Qt::WA_Hover);
   m_label_font.setPixelSize(scale_height(10));
+  m_font_metrics = QFontMetrics(m_label_font);
   auto current_time = boost::posix_time::second_clock::local_time();
   auto bottom_right = ChartPoint(ChartValue(current_time),
     ChartValue(Nexus::Money(10)));
   auto top_left = ChartPoint(
     ChartValue(current_time - boost::posix_time::hours(1)),
     ChartValue(Nexus::Money(10.10)));
   set_region(top_left, bottom_right);
+  setCursor(Qt::BlankCursor);
+  m_dashed_line_pen.setDashPattern({static_cast<double>(scale_width(4)),
+    static_cast<double>(scale_width(4))});
 }
 
 ChartPoint ChartView::convert_pixels_to_chart(const QPoint& point) const {
-  return {};
+  return ChartPoint(
+    map_to(static_cast<double>(point.x()), 0.0,
+      static_cast<double>(m_x_origin), m_top_left.m_x, m_bottom_right.m_x),
+    map_to(static_cast<double>(point.y()), static_cast<double>(m_y_origin),
+      0.0, m_bottom_right.m_y, m_top_left.m_y));
 }
 
 QPoint ChartView::convert_chart_to_pixels(const ChartPoint& point) const {
-  return {};
+  return QPoint(
+    map_to(point.m_x, m_top_left.m_x, m_bottom_right.m_y, 0.0,
+      static_cast<double>(m_x_origin)),
+    map_to(point.m_y, m_bottom_right.m_y, m_top_left.m_y,
+      static_cast<double>(m_y_origin), 0.0));
 }
 
 void ChartView::set_crosshair(const ChartPoint& position) {
+  m_crosshair_pos = convert_chart_to_pixels(position);
```

Comment:
This is a bug. As a general principle you should attempt to implement one function in terms of another function to avoid subtle issues from popping up.